docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,453 @@
+parts of the parameter tree with different optimizers:
+
+```python
+biases_traversal = flax.optim.ModelParamTraversal(
+    lambda path, _: path.endswith('/bias'))
+not_biases_traversal = flax.optim.ModelParamTraversal(
+    lambda path, _: not path.endswith('/bias'))
+
+optimizer_def = flax.optim.MultiOptimizer(
+    (biases_traversal, flax.optim.GradientDescent(learning_rate=0.1)),
+    (not_biases_traversal, flax.optim.GradientDescent(learning_rate=0.05)),
+)
+```
+
+Note how we first define a traversal that selects parameters based on their
+path (which is the concatenation of module scopes and variable name), and then
+create a `MultiOptimizer` that binds a different optimizer for each of these
+separate traversals.
+
+Optax has recently implemented `optax.masked()` that can be used for specifying
+gradient transformations that only applied to a subset of the gradients:
+
+```python
+def dict_map_with_path(fn, d, path=''):  # a crude "ModelParamTraversal"
+    if isinstance(d, dict):
+        return {k: dict_map_with_path(fn, v, path+'/'+k) for k, v in d.items()}
+    return fn(path, d)
+
+is_bias_mask = dict_map_with_path(
+    lambda path, value: path.endswith('/bias'), variables['params'])
+not_bias_mask = jax.tree_map(lambda x: not x, is_bias_mask)
+
+tx = optax.chain(
+    optax.masked(optax.sgd(learning_rate=0.1), mask=is_bias_mask),
+    optax.masked(optax.sgd(learning_rate=0.05), mask=not_bias_mask),
+)
+```
+
+Remarks:
+
+- Above `dict_map_with_path()` emulates `flax.optim.ModelParamTraversal`.
+- Note that now we need `variables['params']` already when the gradient
+  transformation is defined, i.e. before calling `tx.init()`.
+- Note the change in behavior when a parameter matches multiple masks: With
+  `flax.optim.ModelParamTraversal` all but the last match would be discarded.
+  With `optax.chain(optax.masked())` the gradient updates are simply chained
+  and if a gradient matches multiple masks then all those transformations are
+  applied.
+
+## Train State
+[Train State]: #train-state
+
+In Flax it is common to hand around a `TrainState` object that can then be
+used for checkpointing. This simplifies the above [Optax training step] a bit by
+reducing the number of arguments and getting rid of the `static_argnums`.
+
+```python
+class TrainState(flax.struct.PyTreeNode):
+
+  apply_fn: Callable = flax.struct.field(pytree_node=False)
+  variables: flax.core.FrozenDict[str, Any]
+  tx: optax.GradientTransformation = flax.struct.field(pytree_node=False)
+  opt_state: optax.OptState
+```
+
+Users can subclass this state and add more fields:
+
+```python
+class MyTrainState(TrainState):
+  step: int
+```
+
+And then [Optax Training Step] becomes:
+
+```python
+@jax.jit
+def train_step(state, inputs, labels):
+
+  def loss_fn(params):
+    outputs, new_model_state = state.apply_fn(
+      state.variables.copy({'params': params}), inputs, mutable=['batch_stats'])
```

Comment:
nit: the downside of storing variables is that it's easy to accidentally use `state.variables` here which would lead to 0 gradients.

Reply:
Agreed - split "params" from variables.
(undoing https://github.com/google/flax/pull/1011#discussion_r588727155)